File: flappy.py
""""
This is a simple game in python 
that is similar to the popular game 'flappy bird'
"""

import pygame
import random
import os

# Initialize Pygame
pygame.init()

# Screen dimensions
SCREEN_WIDTH = 400
SCREEN_HEIGHT = 600

# Colors
WHITE = (255, 255, 255)
BLACK = (0, 0, 0)
GREEN = (0, 255, 0)

# Game settings
GRAVITY = 0.5
FLAP_STRENGTH = -10
PIPE_WIDTH = 70
PIPE_HEIGHT = 500
PIPE_GAP = 150
PIPE_SPEED = 3

# Load images
BIRD_IMAGES = [pygame.image.load(f"bird_frame_{i}.png") for i in range(3)]
PIPE_IMAGE = pygame.image.load("pipe_detailed.png")
BACKGROUND_IMAGE = pygame.image.load("background_detailed.png")

# Load sounds
FLAP_SOUND = pygame.mixer.Sound("flap.wav")
SCORE_SOUND = pygame.mixer.Sound("score.wav")
HIT_SOUND = pygame.mixer.Sound("hit.wav")

# Bird class
class Bird:
    def __init__(self):
        self.images = BIRD_IMAGES
        self.image_index = 0
        self.image = self.images[self.image_index]
        self.rect = self.image.get_rect()
        self.rect.center = (SCREEN_WIDTH // 4, SCREEN_HEIGHT // 2)
        self.velocity = 0
        self.animation_counter = 0

    def update(self):
        self.velocity += GRAVITY
        self.rect.y += self.velocity
        self.animation_counter += 1
        if self.animation_counter % 5 == 0:
            self.image_index = (self.image_index + 1) % len(self.images)
            self.image = self.images[self.image_index]

    def flap(self):
        self.velocity = FLAP_STRENGTH
        FLAP_SOUND.play()

    def draw(self, screen):
        screen.blit(self.image, self.rect)

    def get_rect(self):
        return self.rect

# Pipe class
class Pipe:
    def __init__(self, x):
        self.image = PIPE_IMAGE
        self.rect_top = self.image.get_rect()
        self.rect_bottom = self.image.get_rect()
        self.rect_top.x = x
        self.rect_bottom.x = x
        self.rect_top.y = random.randint(-PIPE_HEIGHT + PIPE_GAP, 0)
        self.rect_bottom.y = self.rect_top.y + PIPE_HEIGHT + PIPE_GAP

    def update(self):
        self.rect_top.x -= PIPE_SPEED
        self.rect_bottom.x -= PIPE_SPEED

    def draw(self, screen):
        screen.blit(self.image, self.rect_top)
        screen.blit(self.image, self.rect_bottom)

    def get_rects(self):
        return self.rect_top, self.rect_bottom

# Function to display the start menu
def display_start_menu(screen, font):
    screen.fill(WHITE)
    title_text = font.render("Flappy Bird", True, BLACK)
    start_text = font.render("Press SPACE to Start", True, BLACK)
    screen.blit(title_text, (SCREEN_WIDTH // 2 - title_text.get_width() // 2, SCREEN_HEIGHT // 3))
    screen.blit(start_text, (SCREEN_WIDTH // 2 - start_text.get_width() // 2, SCREEN_HEIGHT // 2))
    pygame.display.flip()

# Function to display the scoreboard
def display_scoreboard(screen, font, score):
    screen.fill(WHITE)
    score_text = font.render(f'Your Score: {score}', True, BLACK)
    restart_text = font.render("Press SPACE to Restart", True, BLACK)
    screen.blit(score_text, (SCREEN_WIDTH // 2 - score_text.get_width() // 2, SCREEN_HEIGHT // 3))
    screen.blit(restart_text, (SCREEN_WIDTH // 2 - restart_text.get_width() // 2, SCREEN_HEIGHT // 2))
    pygame.display.flip()

# Main game function
def main():
    screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
    pygame.display.set_caption("Flappy Bird")
    clock = pygame.time.Clock()
    font = pygame.font.SysFont(None, 36)

    # Display start menu
    display_start_menu(screen, font)
    waiting = True
    while waiting:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                return
            if event.type == pygame.KEYDOWN:
                if event.key == pygame.K_SPACE:
                    waiting = False

    bird = Bird()
    pipes = [Pipe(SCREEN_WIDTH + i * (PIPE_WIDTH + 200)) for i in range(3)]
    score = 0

    running = True
    while running:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                running = False
            if event.type == pygame.KEYDOWN:
                if event.key == pygame.K_SPACE:
                    bird.flap()

        bird.update()
        for pipe in pipes:
            pipe.update()
            if pipe.rect_top.right < 0:
                pipes.remove(pipe)
                pipes.append(Pipe(SCREEN_WIDTH + PIPE_WIDTH))
                score += 1
                SCORE_SOUND.play()

        # Check for collisions
        for pipe in pipes:
            if bird.get_rect().colliderect(pipe.get_rects()[0]) or bird.get_rect().colliderect(pipe.get_rects()[1]):
                HIT_SOUND.play()
                running = False

        screen.blit(BACKGROUND_IMAGE, (0, 0))
        bird.draw(screen)
        for pipe in pipes:
            pipe.draw(screen)

        # Draw score
        score_text = font.render(f'Score: {score}', True, BLACK)
        screen.blit(score_text, (10, 10))

        pygame.display.flip()
        clock.tick(30)

    # Display scoreboard
    display_scoreboard(screen, font, score)
    waiting = True
    while waiting:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                return
            if event.type == pygame.KEYDOWN:
                if event.key == pygame.K_SPACE:
                    main()
                    return

    pygame.quit()

if __name__ == "__main__":
    main()

File: flappy.js
const canvas = document.getElementById('gameCanvas');
const ctx = canvas.getContext('2d');

// Resize canvas to fit the window
canvas.width = window.innerWidth;
canvas.height = window.innerHeight;

// Screen dimensions
const SCREEN_WIDTH = canvas.width;
const SCREEN_HEIGHT = canvas.height;

// Colors
const WHITE = '#FFFFFF';
const BLACK = '#000000';

// Game settings
const GRAVITY = 0.5;
const FLAP_STRENGTH = -10;
const PIPE_WIDTH = 70;
const PIPE_HEIGHT = 500;
const PIPE_GAP = 150;
const PIPE_SPEED = 3;
const FRAME_RATE = 30; // 30 frames per second
const FRAME_INTERVAL = 1000 / FRAME_RATE;
const BIRD_SCALE = 2; // Scale factor for the bird

// Load images
const BIRD_IMAGE = new Image();
BIRD_IMAGE.src = 'bird.png';
const PIPE_IMAGE = new Image();
PIPE_IMAGE.src = 'pipe.png';
const BACKGROUND_IMAGE = new Image();
BACKGROUND_IMAGE.src = 'background.png';

// Bird class
class Bird {
    constructor() {
        this.image = BIRD_IMAGE;
        this.x = SCREEN_WIDTH / 4;
        this.y = SCREEN_HEIGHT / 2;
        this.width = this.image.width * BIRD_SCALE;
        this.height = this.image.height * BIRD_SCALE;
        this.velocity = 0;
    }

    update() {
        this.velocity += GRAVITY;
        this.y += this.velocity;
    }

    flap() {
        this.velocity = FLAP_STRENGTH;
    }

    draw() {
        ctx.drawImage(this.image, this.x, this.y, this.width, this.height);
    }

    getRect() {
        return { x: this.x, y: this.y, width: this.width, height: this.height };
    }
}

// Pipe class
class Pipe {
    constructor(x) {
        this.image = PIPE_IMAGE;
        this.x = x;
        this.topY = Math.random() * (-PIPE_HEIGHT + PIPE_GAP);
        this.bottomY = this.topY + PIPE_HEIGHT + PIPE_GAP;
    }

    update() {
        this.x -= PIPE_SPEED;
    }

    draw() {
        ctx.drawImage(this.image, this.x, this.topY);
        ctx.drawImage(this.image, this.x, this.bottomY);
    }

    getRects() {
        return [
            { x: this.x, y: this.topY, width: PIPE_WIDTH, height: PIPE_HEIGHT },
            { x: this.x, y: this.bottomY, width: PIPE_WIDTH, height: PIPE_HEIGHT }
        ];
    }
}

// Function to display the start menu
function displayStartMenu() {
    document.getElementById('startMenu').style.display = 'block';
    document.getElementById('scoreboard').style.display = 'none';
}

// Function to display the scoreboard
function displayScoreboard(score) {
    document.getElementById('scoreText').innerText = `Your Score: ${score}`;
    document.getElementById('startMenu').style.display = 'none';
    document.getElementById('scoreboard').style.display = 'block';
}

// Main game function
function main() {
    let bird = new Bird();
    let pipes = Array.from({ length: 3 }, (_, i) => new Pipe(SCREEN_WIDTH + i * (PIPE_WIDTH + 200)));
    let score = 0;
    let running = true;
    let lastTime = 0;

    function gameLoop(timestamp) {
        if (!running) return;

        const deltaTime = timestamp - lastTime;
        if (deltaTime < FRAME_INTERVAL) {
            requestAnimationFrame(gameLoop);
            return;
        }
        lastTime = timestamp;

        ctx.drawImage(BACKGROUND_IMAGE, 0, 0, SCREEN_WIDTH, SCREEN_HEIGHT);
        bird.update();
        bird.draw();

        for (let pipe of pipes) {
            pipe.update();
            pipe.draw();
            if (pipe.x + PIPE_WIDTH < 0) {
                pipes.shift();
                pipes.push(new Pipe(SCREEN_WIDTH + PIPE_WIDTH));
                score += 1;
            }
        }

        // Check for collisions
        for (let pipe of pipes) {
            if (checkCollision(bird.getRect(), pipe.getRects()[0]) || checkCollision(bird.getRect(), pipe.getRects()[1])) {
                running = false;
                displayScoreboard(score);
                gameStarted = false; // Allow game restart
                return;
            }
        }

        // Draw score
        ctx.fillStyle = BLACK;
        ctx.font = '24px Arial';
        ctx.fillText(`Score: ${score}`, 10, 30);

        requestAnimationFrame(gameLoop);
    }

    function checkCollision(rect1, rect2) {
        return rect1.x < rect2.x + rect2.width &&
               rect1.x + rect1.width > rect2.x &&
               rect1.y < rect2.y + rect2.height &&
               rect1.y + rect1.height > rect2.y;
    }

    document.addEventListener('keydown', (event) => {
        if (event.code === 'Space') {
            bird.flap();
        }
    });

    requestAnimationFrame(gameLoop);
}

// Start the game only once
let gameStarted = false;
document.addEventListener('keydown', (event) => {
    if (event.code === 'Space' && !gameStarted) {
        gameStarted = true;
        document.getElementById('startMenu').style.display = 'none';
        main();
    }
});

// Restart the game after it ends
document.addEventListener('keydown', (event) => {
    if (event.code === 'Space' && !gameStarted) {
        gameStarted = true;
        document.getElementById('scoreboard').style.display = 'none';
        main();
    }
});

displayStartMenu();
